packages/encrypted_storage/lib/src/storage.dart

```diff
@@ -0,0 +1,268 @@
+  }
+
+  FutureOr<void> _onUpgrade(Database db, int oldVersion, int newVersion) {
+    _log
+      ..finest('database upgraded from $oldVersion to $newVersion')
+      ..warning('no upgrade migrations found');
+  }
+
+  FutureOr<void> _onDowngrade(Database db, int oldVersion, int newVersion) {
+    _log
+      ..finest('database downgraded from $oldVersion to $newVersion')
+      ..warning('no downgrade migrations found');
+  }
+
+  /// Clear storage: all records
+  Future<void> clearAll() async {
+    const query = '''
+      DELETE FROM storage;
+    ''';
+
+    await _database.execute(query);
+
+    _log.finest('storage cleared');
+  }
+
+  /// Clear storage: all records in one domain
+  Future<void> clearDomain([String? domain = defaultDomain]) async {
+    final query = '''
+      DELETE FROM storage WHERE domain = "$domain";
+    ''';
+
+    await _database.execute(query);
+
+    _log.finest('domain $domain cleared');
+  }
+
+  /// Write the key-value pair. [value] will be written for the [key] in
+  /// [domain].
+  /// If the pair was already existed it will be overwritten if [overwrite]
+  /// is true (by default)
+  Future<void> set(
+    String key,
+    StorageValue value, {
+    String domain = defaultDomain,
+    bool overwrite = true,
+  }) async {
+    return setDomain(
+      {
+        key: value,
+      },
+      domain: domain,
+      overwrite: overwrite,
+    );
+  }
+
+  /// Write the key-value pair map. [pairs] will be written in [domain].
+  /// If the pair was already existed it will be overwritten if [overwrite]
+  /// is true (by default). Unspecified in [pairs] in db will not be altered
+  /// or deleted.
+  Future<void> setDomain(
+    Map<String, StorageValue> pairs, {
+    String domain = defaultDomain,
+    bool overwrite = true,
+  }) async {
+    if (pairs.isEmpty) {
+      _log.info('setAll called with empty pair map');
+      return;
+    }
+
+    var isFirst = true;
+    final values = pairs.entries.fold('', (previousValue, pair) {
+      final prefix = isFirst ? '' : ', ';
+      final result =
+          // ignore: lines_longer_than_80_chars
+          '$previousValue$prefix("$domain", "${pair.key}", "${pair.value.value}", "${pair.value.iv}" )';
+      isFirst = false;
+      return result;
+    });
+
+    final conflictClause = overwrite ? 'REPLACE' : 'IGNORE';
+    final query = '''
+      INSERT OR $conflictClause INTO storage (domain, key, value, iv) VALUES $values;
+    ''';
+
+    await _database.execute(query);
+  }
+
+  /// Delete by [key] from [domain].
+  Future<void> delete(
+    String key, {
+    String domain = defaultDomain,
+  }) async {
+    return deleteDomain([key], domain: domain);
+  }
+
+  /// Delete by [keys] from [domain].
+  Future<void> deleteDomain(
+    List<String> keys, {
+    String domain = defaultDomain,
+  }) async {
+    if (keys.isEmpty) {
+      _log.info('deleteDomain called with empty key list');
+      return;
+    }
+
+    if (keys.length > 512) {
```

Comment:
What is it related to?

Reply:
sqlite doesn't like long queries with WHERE ... AND ( ... OR ... OR ...) clauses. I've tested with 512.